### PR TITLE
Adds good binary type descriptors

### PR DESCRIPTION
**Pull request by jonwilsdon**
*Description of changes:*

Adds good files for each TL pair and a README that describes the files. As much as possible, I tried to make the representation to be made up of invalid TL pairs. That way even if the implementation is not trying to read the values, it should still fail if it is reading the length incorrectly. The README is intended to be a resource I would have liked to have when starting to understand the binary format.

I split the T7 (Symbol) file into two, for implementations that do not want to implement supporting really large VarUInts. (ion-c, ion-java, ion-js all have problems with `T7-large.10n`).

Below are the issues that I noticed when testing it with the various language implementations, the first three are the most important to take a look at:

- ion-python differs on `5D`, `5E` in `T5.10n` compared to ion-c, ion-js, ion-java (but is fine on the other large numbers)
- ion-python (**Edit:** originally this said ion-java) disagrees with ion-c, ion-js, and ion-java on the value of 62, 63, 64, and 65 in `T6.10n`. (`0097-01T` vs `0096-12T`)
- While creating these files, I noticed that ion-js seems to accept timestamps that it shouldn't (ones that other implementations don't accept and the issue seems to be around the hour and minute requirements). I will follow up on this and create an issue. (**Edit:** Issue https://github.com/amzn/ion-js/issues/534 created)
- ion-c fails reading `T6.10n` (on the larger timestamps)
- ion-c, ion-java, and ion-js fail reading `T7-large.10n`
- ion-js does not support SID0 (Used in `T7-small.10n` and `T7-large.10n`)

If there is interest, I can create a set of bad files for the invalid types as well.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: iontestdata/good/typecodes/README.md [hunk 1]
```diff
@@ -0,0 +1,1153 @@
...
+> 6 byte timestamp with at most VarInt offset, VarUInt year, VarUInt month, VarUInt day, VarUInt hour, and VarUInt minute components  
+> second, fraction_exponent, and fraction_coefficient components are not able to be specified  
+> _0097-01-01T00:28-00:33_
+
+> `67 E1 E1 81 81 81 81 81`  
+> 7 byte timestamp with at most VarInt offset, VarUInt year, VarUInt month, VarUInt day, VarUInt hour, VarUInt minute, and VarUInt second  components  
+> fraction_exponent, and fraction_coefficient components are not able to be specified  
+> _0097-01-01T00:28:01-00:33_
+
+> `68 E1 E1 81 81 81 81 81 E1`
```
Comment (tgregg) [inline iontestdata/good/typecodes/README.md]: I recommend splitting the timestamps with large fractional precision into a separate file like you did with symbols. Some implementations have open issues for supporting arbitrary fractional precision.

Alternatively, you could try just using a lower precision that would work for all implementations, as we already have tests for high fractional precision.

File: iontestdata/good/typecodes/README.md [hunk 2]
```diff
@@ -0,0 +1,1153 @@
...
+
+> `E0 01 00 EA`  
+> Binary Version Marker (BVM)
+
+> `50`  
+> _0._ or _0d0_
+
+> `51 FF`  
+> 1 byte decimal  
+> _0d-63_ or _0E-63_
```
[the remark this answers is not in the document]
Comment (jonwilsdon) [reply iontestdata/good/typecodes/README.md]: That output was taken from ion-js. If 'E' is not valid, does it need to change in ion-js?

The spot where the 'E' is added for IonDecimal: https://github.com/amzn/ion-js/blob/master/src/IonDecimal.ts#L227.

Although it does look like that is only for `.toString()` as it is written out with a 'd' in the IonTextWriter: https://github.com/amzn/ion-js/blob/master/src/IonTextWriter.ts#L166

I'll remove the 'E' variants for Decimal.

File: iontestdata/good/typecodes/README.md [hunk 3]
```diff
@@ -0,0 +1,1153 @@
...
+81 81 81 81 E1 12 12 12 12 12 6E 8E E1 E1 81 81 
+81 81 81 E1 12 12 12 12 12 12 6F  
+ 
+```
+
+> `E0 01 00 EA`  
+> Binary Version Marker (BVM)
+
+> `62 E1 E1`  
+> 2 byte timestamp with at most VarInt offset and VarUInt year components
```
[the remark this answers is not in the document]
Comment (jonwilsdon) [reply iontestdata/good/typecodes/README.md]: The way you rephrased it makes sense to me.

I will change these to `0097T` as it is the most common output. Also, I was wrong in the initial description, it is ion-python that outputs `0096T` and not ion-java.

File: iontestdata/good/typecodes/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgregg) [reply iontestdata/good/typecodes/README.md]: Yeah I don't think there's a good reason to have `IonDecimal.toString()` return the text representation of an Ion float.

File: iontestdata/good/typecodes/README.md (hunk 1 above)
Comment (jonwilsdon) [reply iontestdata/good/typecodes/README.md]: I split the file. Line 5 (now updated) attempts to explain the reasoning why I am sticking with a high fractional precision.

File: iontestdata/good/typecodes/README.md [hunk 4]
```diff
@@ -0,0 +1,1153 @@
+# Type Codes
+
+The files in this folder are used to test all of the good type descriptors for Ion 1.0.  
+Each file contains a valid Ion [Value Stream](http://amzn.github.io/ion-docs/docs/binary.html#value-streams).  
+Where possible, all representations are made up of invalid type descriptors to ensure readers are reading the proper lengths without relying on an implementation for reading the representation. Symbols are an exception; writing a symbol table with a large `max_id` adds unwanted complexity to the tests.
```
[the remark this answers is not in the document]
Comment (jonwilsdon) [reply iontestdata/good/typecodes/README.md]: I updated the line. Let me know if it still isn't clear.

**Comment by jonwilsdon** [comment]
I added a commit that addresses the comments.

File: iontestdata/good/typecodes/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonwilsdon) [reply iontestdata/good/typecodes/README.md]: I created an issue to fix it: https://github.com/amzn/ion-js/issues/536

**Comment by jonwilsdon** [comment]
I noticed while putting together the set of bad files that I had missed listing `1E` as an invalid type descriptor. I added it with the latest commit.

File: iontestdata/good/typecodes/README.md [hunk 5]
```diff
@@ -0,0 +1,1153 @@
...
+ 
+```
+
+> `E0 01 00 EA`  
+> Binary Version Marker (BVM)
+
+> `62 E1 E1`  
+> 2 byte timestamp with at most VarInt offset and VarUInt year components  
+> month, day, hour, minute, second, fraction_exponent, and fraction_coefficient components are not able to be specified  
+> _0096T_
```
[the remark this answers is not in the document]
Comment (tgregg) [reply iontestdata/good/typecodes/README.md]: We are going to raise an error in the cases discussed above and clarify the spec.

**Comment by jonwilsdon** [comment]
Timestamp changes made, squashing and merging.

**Pull request merged**